Data Provider tasks

annavolovyk opened the pull request.

annavolovyk committed:
Data Provider tasks

annavolovyk committed:
Data Provider tasks (updated)

LanaSvit commented on src/test/java/manandwomantests/ManTest.java:
```diff
@@ -0,0 +1,26 @@
+        Assert.assertEquals(ageNow, canRetire);
+    }
+
+    @DataProvider(name = "mantests")
```
DataProvider прийнято виности в окремий клас

LanaSvit commented on src/test/java/manandwomantests/ManTest.java:
```diff
@@ -0,0 +1,26 @@
+@Test(dataProvider = "mantests", description = "Verifying the age for retirement", priority = 10)
+    public void testIsRetired(int age, boolean canRetire) {
+        man.setAge(age);
+        boolean ageNow = man.isRetired();
```
можна не виносити в окрему змінну ageNow, оскільки вона більше ніде не буде використовуватися

LanaSvit commented on src/test/java/manandwomantests/ManTest.java:
```diff
@@ -0,0 +1,26 @@
+    }
+@Test(dataProvider = "mantests", description = "Verifying the age for retirement", priority = 10)
+    public void testIsRetired(int age, boolean canRetire) {
+        man.setAge(age);
```
як на мене, зайва строка, age і так буде братися з провайдера

LanaSvit commented on src/test/java/manandwomantests/PersonTest.java:
```diff
@@ -0,0 +1,112 @@
+    @Test(dataProvider = "manAge")
+    public void testManGetAge(int manActualAge, int manExpectedAge) {
+        person.setAge(manActualAge);
+        Assert.assertEquals(person.getAge(), manExpectedAge);
```
звідки тут manExpectedAge?

LanaSvit commented on src/test/java/personTests/PersonTest.java:
```diff
@@ -0,0 +1,111 @@
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+public class PersonTest {
```
вже був клас PersonTest

annavolovyk committed:
Data Provider tasks (updated2)

annavolovyk commented:
@LanaSvit Good evening! Could you please look at this commit e299c65416837dfeb1835a8704596c2d285a6f88
1) added separate package and class for Data Provider
2) re-wrote some of the @Tests + added to them TestDataProvider class as a parameter + re-wrote data providers
3) did not write Data Providers for two methods in WomanTest (register and deregister methods, as in the previous h/a used variable null for 'partner' (if it is needed will try to re-write methods (although actually have difficulty with that => if you could help me with those two methods, their @Tests and Data Providers for them, I would be very grateful :)
4) added .idea to gitignore
5) and here is a screenshot of the successful build https://prnt.sc/mhzo17CEeuDX

annavolovyk commented:
@LanaSvit І тільки прочитала попередній коментар до д/з. Я звичайно всі коментарі намагаюся доробити і дороблятиму, тому що мені важливо, насамперед, зрозуміти, що у мене зроблено "криво". 
Але, на жаль, зараз етап пленнінгів на моєму проекті, а це мітинги до 23:00 (тому що американський офіс) і я можу сісти апдейтити свої д/з тільки пізно увечері, тому вибачте, що змушую Вас чекати :(

annavolovyk commented:
@LanaSvit Світлано, добрий день! Сподіваюся, що лінки тут працюють нормально і Ви змогли побачити мій апдейт в комміті і повідомлення вище ((. В системі https://my-ua.robotdreams.cc вже не могла залишати коментарі.

LanaSvit commented:
Мені не подобається, що в кожному тесті викликаються сетери. Такий підхід ускладнює дебаг тестів у випадку їх падіння. Адже буде не відомо, тест впав через проблему в методі, що тестується, чи через сетер. З цієї ж причини можуть бути false positive рузультати, тобто тест пройте, бо сетиться відразу необхідне значення, хоча сам метод буде не робочий. Тому я би йшла в бік того, щоб відразу передавати дані з провайдера, необхідні для створення потрібного об'єкта.
Наприклад
@DataProvider(name = "manRetiredData")
  public static Object[][] manRetiredData() {
      Man man1 = new Man("Oleksandr", "Petrov", "man", 25, null);
      Man man2 = new Man("Nick", "Logay", "man", 66, null);
      Man man3 = new Man("Petr", "Koval", "man", 65, null);
      return new Object[][]{
              {man1, false},
              {man2, true},
              {man3, true}
      };
  }

  @Test(dataProvider = "manRetiredData", dataProviderClass = Data.class)
  public void testSetRetired(Man man, boolean expectedIsRetiredResult) {
      Assert.assertEquals(man.isRetired(), expectedIsRetiredResult, "Problems with retired setter method");
  }

LanaSvit commented on src/main/java/manandwoman/Woman.java:
```diff
@@ -0,0 +1,40 @@
+
+    public Object registerPartnership(Person partner) {
+        if (getPartner() == null) {
+            setLastName(String.valueOf(getPartner().getLastName()));
```
можна без String.valueOf, бо getLastName повертає String

LanaSvit commented on src/main/java/manandwoman/Woman.java:
```diff
@@ -0,0 +1,40 @@
+
+
+    public Object registerPartnership(Person partner) {
+        if (getPartner() == null) {
```
тут логічна помилка. Якщо getPartner() == null, то як можна у null викликати getLastName()?
getLastName() потрібно викликати у partner, якого ви передаєте як параметр в метод

LanaSvit commented on src/main/java/manandwoman/Woman.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+
+    public Object registerPartnership(Person partner) {
```
навіщо повертати getPartner() в цьому методі? краще нічого не повертати, як на мене

LanaSvit commented on src/main/java/manandwoman/Woman.java:
```diff
@@ -0,0 +1,40 @@
+            System.out.println(getFirstName() + " " + getLastName() + "- has no partner.");
+            return;
+        }
+        if (returnToOldLastName && oldLastName != null) {
```
потрібно зробити так, щоб oldLastName не міг бути null, це логічно не правильно, що у жінки відсутнє дівоче прізвище

annavolovyk committed:
Data Provider tasks (updated3)

annavolovyk commented:
@LanaSvit hi! When you have time please have a look at this commit: 
f49f7f0dae944325aabb42f3615795af474b2ccf

LanaSvit commented:
@annavolovyk Added new comments

annavolovyk committed:
Data Provider tasks (updated4)

annavolovyk commented:
@LanaSvit Добрий вечір! Ось оновлений коміт a488297e47f419de364c72abb25a3fee1c457e02 Дякую!

Світлано, у мене до Вас також ще одне прохання є, а саме пояснити мені, як задавати партнера в конструкторі. Я в своїх тестах перевярла, що partner = null через те, що так і не змогла задати його. Тобто, ми в класі Person створили поле private Person partner і передали в ньому об'єкт. І коли ми в констукторі намагаємося ініціалізувати його, мені видає помилку, що я не тим значенням його ініціалізую (як тут на скріні https://prnt.sc/dPOhAqBKd4xU). Вимагає змінити об'єкт на стрінгу. Тому і питання, за допомогою чого нам задати значення в partner? Заздалегідь вдячна за пояснення!  І сподіваюся, що я Вас не заплутала)

LanaSvit commented:
В класі Person створили поле private Person partner, тобто це має бути об'єкт partner типу Person. Тому, коли в класі Man, що наслідується, ви намагаєтеся передати в конструктор стрінг, отримуєте цю помилку, тому що тип стрінг і тип Person не співпадають. Отже, вам потрібно вирішити, якого типу змінна має бути partner: Person чи string? Судячи зі скріншота, ви хочете, щоб в partner можна було передавати лише ім'я, тоді і в класі Person має бути сказано, що partner має бути типу стрінг. Або, якщо все ж вам потрібно, щоб partner був об'єктом типу Person, тоді і в дочірніх класах в конструкторі передавайте partner як об'єктом типу Person (напр, {firstname, lastname, age, partner})
Так зрозуміліше?

annavolovyk commented:
@LanaSvit так, дякую, я розумію логіку, чому так має працювати, і мені якраз String і не потрібен. Я не можу зрозуміти, як саме передати partner об'єктом типу Person. Можна Вас попросити модифікувати цей приклад, щоб я зрозуміла, як передати цьому чоловіку існуючого партнера:  public Man(String firstname, String lastname, **Person partner,** int age) {
        super("Alex", "Smith", **null** , 65); }

LanaSvit commented:
Цей приклад мені не дуже зрозумілий. Ви хочете в конструктор передавати вже готові значення? У вас же правильно написаний конструктор був
    public Man(String firstname, String lastname, int age, Person partner) {
        super(firstname, lastname, partner, age);
    }

LanaSvit commented:
Давайте я вам покажу, як би я зробила тест womanRegisterPartnership

    @DataProvider(name = "womanRegisterPartnership")
    public Object[][] womanRegisterPartnership(){
        return new Object[][] {{"Mary", "Black", 45, null},
                {"Anna", "Stabile", 15, null}
        };
    }

    @Test(dataProvider = "womanRegisterPartnership", dataProviderClass = TestDataProvider.class)
    public void registerPartnership(String firstName, String lastName, int age, Person partner) {
        woman = new Woman(firstName, lastName, age, partner);
        partner1 = new Man("F", "L", 44, null) ;
        woman.registerPartnership(partner1);
        Assert.assertEquals(woman.getPartner(), partner1, "Has a partner");
    }
Тобто, щоб перевірити, чи працює метод registerPartnership, мені потрібно створити об'єкт жінки без партнера (я роблю відразу два таких в дата провайдері), у тесті викликати метод registerPartnership, передавши в нього параметром об'єкт типу Man (наслідується від Person). В асерта перевіряю, що метод спрацював коректно, тобто що woman.getPartner() відповідає тому об'єкту, який переданий параметром.
